gnd/src/main/java/com/google/android/gnd/ui/map/MapProvider.java

```diff
@@ -16,28 +16,91 @@
 
 package com.google.android.gnd.ui.map;
 
-import android.util.Pair;
-import androidx.fragment.app.Fragment;
+import androidx.annotation.StringRes;
+import com.google.android.gms.maps.GoogleMap;
+import com.google.android.gnd.R;
+import com.google.android.gnd.rx.annotations.Hot;
+import com.google.android.gnd.ui.map.gms.GoogleMapsFragment;
 import com.google.common.collect.ImmutableList;
-import io.reactivex.Single;
+import io.reactivex.Flowable;
+import io.reactivex.processors.FlowableProcessor;
+import io.reactivex.processors.PublishProcessor;
+import java8.util.function.Supplier;
+import javax.inject.Inject;
+import javax.inject.Singleton;
 
 /**
- * Common interface for various map provider libraries.
- *
- * <p>Map Type refers to the basemap shown below map features and offline satellite imagery. It's
- * called "map styles" in Mapbox and "basemaps" in Leaflet.
+ * Creates a new {@link MapFragment}. Currently only {@link GoogleMapsFragment} is present, but the
+ * goal is to choose it based on user's preference.
  */
-public interface MapProvider {
-  void restore(Fragment fragment);
+@Singleton
+public class MapProvider {
+
+  private static final BasemapSource[] BASEMAP_SOURCES = initSources();
+
+  @Hot private final FlowableProcessor<MapAdapter> map = PublishProcessor.create();
+
+  @Inject
+  public MapProvider() {}
+
+  private static BasemapSource[] initSources() {
+    return new BasemapSource[] {
+      new BasemapSource(
+          GoogleMapsFragment::new,
+          ImmutableList.<MapType>builder()
+              .add(new MapType(GoogleMap.MAP_TYPE_NORMAL, R.string.normal))
+              .add(new MapType(GoogleMap.MAP_TYPE_SATELLITE, R.string.satellite))
+              .add(new MapType(GoogleMap.MAP_TYPE_TERRAIN, R.string.terrain))
+              .add(new MapType(GoogleMap.MAP_TYPE_HYBRID, R.string.hybrid))
+              .build())
+    };
+  }
+
+  public static BasemapSource getSource() {
+    // TODO: Select based on user preference.
+    return BASEMAP_SOURCES[0];
+  }
+
+  public MapFragment createFragment() {
+    return getSource().supplier.get();
+  }
+
+  public ImmutableList<MapType> getMapTypes() {
+    return getSource().mapTypes;
+  }
+
+  @Hot
+  public Flowable<MapAdapter> getMapAdapter() {
+    return map;
+  }
+
+  public void setMapAdapter(MapAdapter adapter) {
+    map.onNext(adapter);
+  }
 
-  Fragment getFragment();
+  private static class BasemapSource {
```

Comment:
We're likely going to rename `OfflineBasemapSource` to `BasemapSource` eventually. 

It's no problem since it's private, just something to keep in mind.

Reply:
Do you have any suggestions to what this should be called instead?

Reply:
Sorry just saw this.  Maybe `BasemapType`?